### PR TITLE
Fix some bugs from #1219 Indexing#descendant_uris capable of prioritizing at front of list

### PR DESCRIPTION
**Pull request by jrochkind**

### CODE REVIEW

**Comment by jrochkind** [comment]
Thanks! 

It seems to have failed the travis check. Which shocked me. 

Turns out, it failed rubocops "cyclomatic complexity".  As a result, it didn't even get to running the actual tests, it refused to run them. 

Pretty sure this is safe, but, really rubocop?  Do the rest of y'all really like this sort of thing, rubocop rejecting a PR for 'cyclomatic complexity' cause I added a guard clause to fix a bug? 

Have no idea what you want to do with this one.

**Comment by jeremyf** [comment]
@jrochkind To reduce the Cyclomatic Complexity, extract a private method for the internal of the reindex loop. (This also helps separate layers of abstraction regarding what is happening in a method).

The build is considered broken if Rubocop finds a cyclomatic complexity that is too high.

**Comment by jeremyf** [comment]
Also I added a `.git/hooks/pre-push` (and `chmod +x .git/hooks/pre-push`) to ensure that rubocop runs before push.

```shell
#!/bin/sh
bundle exec rubocop
```

**Comment by jrochkind** [comment]
I tried to extract the inside of `descendants.each do |uri|` to a method, but couldn't figure out any way to do it in a non-insane way. The loop body relies on the arguments to `reindex_everything` as well as the local `batch` variable, would have to pass them all in, seems kind of insane and not an increase in code legibility or maintainability. 

I suppose this is trying to point us to a more significant refactoring where those things are not arguments to a `reindex_everything` method but some kind of object with parameters in a constructor and a no-arg reindex_everything. Maybe. 

Also adding another method means adding another ClassMethod to the ActiveFedora::Indexing concern module (which gets included... somewhere), which I'm loathe to do too, it's too bad these are class methods in the first place, and I'd rather not add more. 

Honestly this feels like bike-shedding with rubocop cyclomatic complexity calculator, a very non-communicative partner at that. 

Not sure what to do here. If these bugfixes are not merged in though, #1219 should probably be reverted, it's buggy.

**Comment by hackmastera** [comment]
So I find @jrochkind's argument convincing and wanted to bump this discussion.

But I also got interested and looked at the method that rubocop was flagging. Here's an idea I had; what do y'all think @jrochkind @jeremyf 

https://github.com/hackmastera/active_fedora/commit/dbc0bb62eec9ff88f92e09c94004d9cb5d6186e2

**Comment by jrochkind** [comment]
Thanks @hackmastera! I personally don't think extracting this stuff to a different class makes the code more readable or maintainable, but if you guys do, and it makes rubocop happy so this can be merged, I'll go along. 

Seriously all I did was add a one-line guard clause and remove a one-line ambiguious `descendants.shift` that was meant to do the same thing but no longer working.  I think I left the code no worse than I found it, I disagree with rubocop's algorithmic analysis that my change is unacceptable. Is there any way to just tell rubocop to ignore it this time?

Either way, I regret that #1219 is in master but will not work, it'll always raise when it gets to the root_url presently, and will skip an arbitrary permissions url, it's not good code and can't be left in master without a fix.

**Comment by hackmastera** [comment]
I think it makes the method slightly more readable but not enough to justify the weight of the solution, really. And I'd prefer not to spend time writing tests for it.

There is a way to tell rubocop to ignore the file for this warning; you would add a line to https://github.com/projecthydra/active_fedora/blob/master/.rubocop.yml#L36

I'll defer to @jeremyf.

**Comment by mjgiarlo** [comment]
This is a good small fix. :+1: to adding an inline rubocop exception around this method.

**Comment by jrochkind** [comment]
I tried telling rubycop to ignore cyclomatic complexity for this file, but it's now complaining about OTHER things that are not part of this PR (but were part of #1219, but got in anyway somehow), and blocking the CI.  I don't completely understand the other things it's complaining about, they don't seem to be problems to me. 

~~~
lib/active_fedora/indexing.rb:90:32: W: Lint/UnusedMethodArgument: Unused method argument - batch_size.
        def reindex_everything(batch_size: 50, softCommit: true, progress_bar: false, final_commit: false)
                               ^^^^^^^^^^
lib/active_fedora/indexing.rb:95:11: W: Lint/UselessAssignment: Useless assignment to variable - progress_bar_controller.
          progress_bar_controller = ProgressBar.create(total: descendants.count, format: "%t: |%B| %p%% %e") if progress_bar
          ^^^^^^^^^^^^^^^^^^^^^^^
~~~

It is not an un-used method argument, it is used;  it is not a useless assignment, the assignment it is used. 

??

**Comment by jrochkind** [comment]
In fact, on my local machine, it is _not_ complaining about cyclomatic complexity anymore, even without telling it to ignore. Very confused in this fight with rubocop.  I just wanna get this one-line bugfix in!

**Comment by jrochkind** [comment]
Disabled cyclomatic complexity check for this single method, using an inline comment rubocop directive (thanks @mjgiarlo for telling me that was a thing). 

@jeremyf , you're currently the reviewer for this I think, so we can't proceed without your input, yes? Can you please check in and tell us what you think, so we can get this recently introduced bug out of master? Thanks!

File: lib/active_fedora/indexing.rb [hunk 1]
```diff
@@ -87,15 +87,17 @@ def index_config
         # @param [Boolean] softCommit - Do we perform a softCommit when we add the to_solr objects to SolrService. Default true.
         # @param [Boolean] progress_bar - If true output progress bar information. Default false.
         # @param [Boolean] final_commit - If true perform a hard commit to the Solr service at the completion of the batch of updates. Default false.
-        def reindex_everything(batch_size: 50, softCommit: true, progress_bar: false, final_commit: false)
+        def reindex_everything(batch_size: 50, softCommit: true, progress_bar: false, final_commit: false) # rubocop:disable Metrics/CyclomaticComplexity
```
Comment (jcoyne) [inline lib/active_fedora/indexing.rb]: I don't think you can scope this to a method.  What I see in the docs is that you want to reinable it at the end of the method: http://rubocop.readthedocs.io/en/latest/configuration/#disabling-cops-within-source-code

Comment (jrochkind) [inline lib/active_fedora/indexing.rb]: Huh, I wonder why it worked then, as a "single line with an end-of-line comment."? But okay, I'll quick change it to the disable/enable pair you are asking for. And then I'm done with this for a while! If this isn't good enough (which might be reasonable), I'd ask someone with commit rights to do me a favor and revert #1219 which is buggy without this.

Comment (jcoyne) [inline lib/active_fedora/indexing.rb]: @jrochkind I think it worked because it turned it off until the end of the file.

**Comment by jeremyf** [comment]
@jcoyne Can you approve or dismiss your review?

**Comment by barmintor** [comment]
@jeremyf I'm working on a suggested change to prevent the codestyle removal.

**Comment by jeremyf** [comment]
I believe this is submsumed by #1230

**Pull request closed without merging**